File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.py
class Solution:
    def reqdays(self, weights, capacity):
        days =1
        load =0
        for i in range(len(weights)):
            if load + weights[i] > capacity:
                days += 1
                load = weights[i]
            else:
                load += weights[i]
        return days

    def shipWithinDays(self, weights: List[int], days: int) -> int:
        low = max(weights)
        high = sum(weights)
        ans = float("inf")
        while low <= high:
            mid = int(low + (high - low) / 2)
            daysneeded = self.reqdays(weights, mid)
            if daysneeded <= days:
                ans = min(ans, mid)
                high = mid - 1
            else:
                low = mid + 1
        return ans

File: 1056-capacity-to-ship-packages-within-d-days/capacity-to-ship-packages-within-d-days.cpp
class Solution {
public:
    int numdays(vector<int>& weights,int capacity){
        int day=1;
        int load=0;
        for(int i=0;i<weights.size();i++){
            if(load+weights[i]>capacity){
                day++;
                load=weights[i];
            }
            else{
                load+=weights[i];
            }
            
        }
        return day;
    }
    int shipWithinDays(vector<int>& weights, int days) {
        int n=weights.size();
        int low=*max_element(weights.begin(),weights.end());
        int sum=0;
        int ans=INT_MAX;
        for(auto it:weights){
            sum+=it;
        }
        int high=sum;
        while(low<=high){
            int mid=low+(high-low)/2;
            int nodays=numdays(weights,mid);
            if(nodays<=days){
                ans=min(ans,mid);
                high=mid-1;
            }
            else{
                low=mid+1;
            }
        }
        return ans;
    }
};